src/Common/src/Interop/Unix/System.Net.Security.Native/Interop.GssApiException.cs

```diff
@@ -0,0 +1,77 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class NetSecurityNative
+    {
+        internal sealed class GssApiException : Exception
+        {
+            private readonly Status _minorStatus;
+
+            public Status MinorStatus
+            {
+                get { return _minorStatus;  }
+            }
+
+            public GssApiException(string message) : base(message)
+            {
+            }
+
+            public GssApiException(Status majorStatus, Status minorStatus)
+                : base(GetGssApiDisplayStatus(majorStatus, minorStatus))
+            {
+                HResult = (int)majorStatus;
+                _minorStatus = minorStatus;
+            }
+
+            public static void ThrowIfError(string message, Status majorStatus, Status minorStatus)
+            {
+                if (majorStatus != Status.GSS_S_COMPLETE)
+                {
+                    GssApiException ex = new GssApiException(majorStatus, minorStatus);
+                    throw ex;
+                }
+            }
+
+            private static string GetGssApiDisplayStatus(Status majorStatus, Status minorStatus)
+            {
+                KeyValuePair<Status, bool>[] statusArr = { new KeyValuePair<Status, bool>(majorStatus, false), new KeyValuePair<Status, bool>(minorStatus, true) };
+                int length = statusArr.Length;
+                string[] msgStrings = new string[length];
```

Comment:
You could probably simplify all of this code and eliminate some allocations in the process (though I realize this is for an exceptional code path) by doing something instead like:
```C#
private static string GetGssApiDisplayStatus(Status majorStatus, Status minorStatus)
{
    string majorString = GetGssApiDisplayStatus(majorStatus, false);
    string minorString = GetGssApiDisplayStatus(minorStatus, true);
    return majorString != null && minorString != null ?
        SR.Format(SR.net_gssapi_operation_failed_detailed, majorString, minorString) :
        SR.Format(SR.net_gssapi_operation_failed, majorStatus.ToString("x8"), minorStatus.ToString("x8");
}

private static unsafe string GetGssApiDisplayStatus(Status status, bool minor)
{
    var displayBuffer = default(GssBuffer);
    Interop.NetSecurityNative.Status minStat;
    try
    {
        Interop.NetSecurityNative.Status displayCallStatus = DisplayStatus(out minStat, status, minor, &displayBuffer);
        return displayCallStatus == Status.GSS_S_COMPLETE ?
            Marshal.PtrToStringAnsi(displayBuffer.data) :
            null;
    }
    finally { displayBuffer.Dispose(); }
}
```
or something like that. Not a high priority, though.